### PR TITLE
[Sprint 2] Feature Logic Product List

### PR DESCRIPTION
**Pull request by tantuan127**
List All Product
Link redmine : https://dev.framgia.com/redmine/issues/113296
Task : Logic List All Product
Affect : mainPage

### CODE REVIEW

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductFragment.java [hunk 1]
```diff
@@ -22,23 +37,30 @@ public static ListProductFragment newInstance() {
...
+                new ProductRemoteDataSource(FOrderServiceClient.getInstance()),
+                new ProductLocalDataSource(realmApi));
+        ListProductContract.Presenter presenter =
+                new ListProductPresenter(mViewModel, productRepository, domainRepository);
+        mViewModel.setPresenter(presenter);
 
         FragmentListproductBinding binding =
                 DataBindingUtil.inflate(inflater, R.layout.fragment_listproduct, container, false);
         binding.setViewModel((ListProductViewModel) mViewModel);
+        binding.setProductList(this);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductFragment.java]: Please remove

File: app/src/main/res/layout/fragment_listproduct.xml [hunk 2]
```diff
@@ -19,6 +23,8 @@
     <android.support.v7.widget.RecyclerView
         android:layout_width="match_parent"
         android:layout_height="match_parent"
+        app:layoutManager="@{LayoutManagers.grid(productList.getResources.getInteger(R.integer.item_col))}"
```
Comment (daolq3012) [inline app/src/main/res/layout/fragment_listproduct.xml]: Please provide it from ViewModel and bind to layout

File: app/src/main/java/com/framgia/forder/data/source/ProductDataSource.java [hunk 3]
```diff
@@ -27,5 +27,7 @@
 
     interface RemoteDataSource {
         Observable<List<Product>> getListProduct(int domainId);
+
+        Observable<List<Product>> getListAllProduct(int domainId);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/ProductDataSource.java]: Please remove. you can use ``Observable<List<Product>> getListProduct(int domainId);``

File: app/src/main/java/com/framgia/forder/utils/binding/BindingUtils.java [hunk 4]
```diff
@@ -5,6 +5,7 @@
 import android.support.v4.app.FragmentPagerAdapter;
 import android.support.v4.view.ViewPager;
 import android.support.v7.widget.AppCompatSpinner;
+import android.support.v7.widget.GridLayoutManager;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/utils/binding/BindingUtils.java]: remove unused import

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductPresenter.java [hunk 5]
```diff
@@ -8,16 +20,55 @@
...
+        mCompositeSubscription.clear();
+        mProductRepository.closeTransaction();
+    }
+
+    @Override
+    public void addToCart(Product product) {
+        if (product == null) {
+            return;
+        }
+        mProductRepository.addToCart(product);
```
Comment (tranminhtri911) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductPresenter.java]: Qua bên mainpage presenter tham khảo add vào card đi, không subscri vào sao mà xử lý (kill)

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java [hunk 6]
```diff
@@ -29,4 +49,36 @@ public void onStop() {
     public void setPresenter(ListProductContract.Presenter presenter) {
         mPresenter = presenter;
     }
+
+    @Override
+    public void onGetListAllProductError(BaseException exception) {
+        //TODO Load list Error!
```
Comment (tranminhtri911) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java]: bỏ todo

File: app/src/main/java/com/framgia/forder/screen/mainpage/MainPagePresenter.java [hunk 7]
```diff
@@ -23,6 +21,8 @@
     private ProductRepository mProductRepository;
     private ShopRepository mShopRepository;
     private DomainRepository mDomainRepository;
+    private final int START_SUB_LIST = 0;
```
Comment (tranminhtri911) [inline app/src/main/java/com/framgia/forder/screen/mainpage/MainPagePresenter.java]: private static final

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java [hunk 8]
```diff
@@ -29,4 +49,45 @@ public void onStop() {
     public void setPresenter(ListProductContract.Presenter presenter) {
         mPresenter = presenter;
     }
+
+    @Override
+    public void onGetListAllProductError(BaseException exception) {
+        mNavigator.showToast(R.string.cannot_load_the_data);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java]: Please change to `` mNavigator.showToast(exception.getMessage());``

**Pull request merged**